File: ios/MitiMaiti/ViewModels/FamilyViewModel.swift
import SwiftUI

@MainActor
class FamilyViewModel: ObservableObject {
    @Published var members: [FamilyMember] = []
    @Published var suggestions: [FamilySuggestion] = []
    @Published var currentInvite: FamilyInvite?
    @Published var isLoading = false
    @Published var error: String?
    @Published var selectedTab = 0

    // New state for web-parity
    @Published var activeTab: Int = 0 // 0 = members, 1 = suggestions
    @Published var selectedMemberId: String? = nil // nil = list view, string = detail view
    @Published var showInviteModal = false
    @Published var showRevokeAllModal = false
    @Published var toastMessage: String? = nil

    private let api = APIService.shared

    // MARK: - Computed

    var selectedMember: FamilyMember? {
        guard let id = selectedMemberId else { return nil }
        return members.first(where: { $0.id == id })
    }

    // MARK: - Data Loading

    func loadFamily() {
        isLoading = true
        Task {
            do {
                let result = try await api.fetchFamily()
                members = result.members
                suggestions = result.suggestions
                isLoading = false
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }

    // MARK: - Invite

    func generateInvite() {
        Task {
            do {
                currentInvite = try await api.generateInvite()
            } catch {
                self.error = error.localizedDescription
            }
        }
    }

    func copyInviteCode() {
        guard let invite = currentInvite else { return }
        UIPasteboard.general.string = invite.code
        showToast("Invite code copied!")
    }

    func shareInviteCode() {
        guard let invite = currentInvite else { return }
        let text = "Join my family circle on MitiMaiti! Use code: \(invite.code)\n\(invite.deepLink)"
        let activityVC = UIActivityViewController(activityItems: [text], applicationActivities: nil)
        if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
           let rootVC = windowScene.windows.first?.rootViewController {
            rootVC.present(activityVC, animated: true)
        }
    }

    // MARK: - Permissions

    func updatePermission(memberId: String, keyPath: WritableKeyPath<FamilyPermissions, Bool>, value: Bool) {
        guard let idx = members.firstIndex(where: { $0.id == memberId }) else { return }
        members[idx].permissions[keyPath: keyPath] = value
    }

    func enableAllPermissions(memberId: String) {
        guard let idx = members.firstIndex(where: { $0.id == memberId }) else { return }
        members[idx].permissions = .allEnabled
        showToast("All permissions enabled for \(members[idx].name)")
    }

    func disableAllPermissions(memberId: String) {
        guard let idx = members.firstIndex(where: { $0.id == memberId }) else { return }
        members[idx].permissions = .allDisabled
        showToast("All permissions disabled for \(members[idx].name)")
    }

    // MARK: - Revoke

    func revokeMember(memberId: String) {
        guard let idx = members.firstIndex(where: { $0.id == memberId }) else { return }
        members[idx].status = .revoked
    }

    func revokeAllMembers() {
        for idx in members.indices {
            members[idx].permissions = .allDisabled
        }
        showToast("All family access revoked")
    }

    // MARK: - Suggestions

    func likeSuggestion(id: String) {
        guard let suggestion = suggestions.first(where: { $0.id == id }) else { return }
        let name = suggestion.suggestedUser.displayName
        suggestions.removeAll { $0.id == id }
        showToast("You liked \(name)!")
    }

    func passSuggestion(id: String) {
        suggestions.removeAll { $0.id == id }
        showToast("Passed")
    }

    // MARK: - Toast

    private func showToast(_ message: String) {
        toastMessage = message
        Task {
            try? await Task.sleep(nanoseconds: 2_500_000_000)
            if toastMessage == message {
                toastMessage = nil
            }
        }
    }
}

File: android/app/src/main/java/com/mitimaiti/app/viewmodels/FamilyViewModel.kt
package com.mitimaiti.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mitimaiti.app.models.*
import com.mitimaiti.app.services.APIService
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class FamilyViewModel : ViewModel() {
    private val _members = MutableStateFlow<List<FamilyMember>>(emptyList())
    val members: StateFlow<List<FamilyMember>> = _members.asStateFlow()
    private val _suggestions = MutableStateFlow<List<FamilySuggestion>>(emptyList())
    val suggestions: StateFlow<List<FamilySuggestion>> = _suggestions.asStateFlow()
    private val _currentInvite = MutableStateFlow<FamilyInvite?>(null)
    val currentInvite: StateFlow<FamilyInvite?> = _currentInvite.asStateFlow()
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    private val _selectedTab = MutableStateFlow(0)
    val selectedTab: StateFlow<Int> = _selectedTab.asStateFlow()
    private val _selectedMemberId = MutableStateFlow<String?>(null)
    val selectedMemberId: StateFlow<String?> = _selectedMemberId.asStateFlow()
    private val _showInviteModal = MutableStateFlow(false)
    val showInviteModal: StateFlow<Boolean> = _showInviteModal.asStateFlow()
    private val _showRevokeAllModal = MutableStateFlow(false)
    val showRevokeAllModal: StateFlow<Boolean> = _showRevokeAllModal.asStateFlow()
    private val _toastMessage = MutableStateFlow<String?>(null)
    val toastMessage: StateFlow<String?> = _toastMessage.asStateFlow()
    val selectedMember: FamilyMember? get() = _members.value.firstOrNull { it.id == _selectedMemberId.value }

    fun selectTab(tab: Int) { _selectedTab.value = tab }
    fun selectMember(id: String?) { _selectedMemberId.value = id }
    fun toggleInviteModal() { _showInviteModal.value = !_showInviteModal.value }
    fun toggleRevokeAllModal() { _showRevokeAllModal.value = !_showRevokeAllModal.value }

    fun loadFamily() { viewModelScope.launch { _isLoading.value = true; APIService.fetchFamily().onSuccess { (m, s) -> _members.value = m; _suggestions.value = s }.onFailure { _error.value = "Failed to load family data" }; _isLoading.value = false } }
    fun generateInvite() { viewModelScope.launch { APIService.generateInvite().onSuccess { _currentInvite.value = it; _showInviteModal.value = true } } }

    fun updatePermission(memberId: String, permissionName: String, value: Boolean) {
        _members.value = _members.value.map { m -> if (m.id == memberId) { val p = m.permissions; m.copy(permissions = when (permissionName) {
            "canViewProfile" -> p.copy(canViewProfile = value); "canViewPhotos" -> p.copy(canViewPhotos = value); "canViewBasics" -> p.copy(canViewBasics = value)
            "canViewSindhi" -> p.copy(canViewSindhi = value); "canViewMatches" -> p.copy(canViewMatches = value); "canSuggest" -> p.copy(canSuggest = value)
            "canViewCulturalScore" -> p.copy(canViewCulturalScore = value); "canViewKundli" -> p.copy(canViewKundli = value); else -> p }) } else m }
        showToast("Permission updated")
    }

    fun enableAllPermissions(memberId: String) { _members.value = _members.value.map { if (it.id == memberId) it.copy(permissions = FamilyPermissions(true, true, true, true, true, true, true, true)) else it }; showToast("All permissions enabled") }
    fun disableAllPermissions(memberId: String) { _members.value = _members.value.map { if (it.id == memberId) it.copy(permissions = FamilyPermissions(false, false, false, false, false, false, false, false)) else it }; showToast("All permissions disabled") }
    fun revokeMember(memberId: String) { _members.value = _members.value.map { if (it.id == memberId) it.copy(status = FamilyMemberStatus.REVOKED) else it }; _selectedMemberId.value = null; showToast("Access revoked") }
    fun revokeAllMembers() { _members.value = _members.value.map { it.copy(status = FamilyMemberStatus.REVOKED) }; _showRevokeAllModal.value = false; showToast("All access revoked") }
    fun likeSuggestion(id: String) { _suggestions.value = _suggestions.value.filter { it.id != id }; showToast("Added to your feed!") }
    fun passSuggestion(id: String) { _suggestions.value = _suggestions.value.filter { it.id != id } }
    private fun showToast(message: String) { _toastMessage.value = message; viewModelScope.launch { delay(2000); _toastMessage.value = null } }
}
